src/anml/data/data.py

```diff
@@ -168,21 +176,51 @@ def process_params(self, df: pd.DataFrame):
         
         design_mat_blocks = []
         constr_mat_blocks = []
+        re_mat_groups = defaultdict(dict)
         lbs = []
         ubs = []
+        self.fe_variables_names = []
         for param_set in self._param_set:
             for parameter in param_set.parameters:
                 for variable in parameter.variables:
-                    design_mat_blocks.append(variable.design_mat(df=df))
+                    design_mat = variable.design_mat(df=df)
+                    design_mat_blocks.append(design_mat)
+                    var_name = parameter.param_name + '_' + variable.covariate
+                    self.fe_variables_names.append(var_name)
+                    if variable.add_re:
+                        re_mat_groups[variable.col_group][var_name] = design_mat
```

Comment:
I feel like here we might over complicate things. Can we just stack the matrix like what we do for the fixed effects here? Since didn't see multiple `col_group`. And then we could turn the matrix into block diagonal?

Reply:
right now each variable can have a different `col_group`, therefore here I am grouping variables with the same `col_group` together. for instance, if two variables have the same `col_group = grouping1`, then I collect their design matrices (which are two column vectors) and stack together (which is `N-by-2`).

Reply:
we could assume only one `col_group` also. I saw a list of `col_group` in `DataSpecs`, so I assumed that we actually wanted that flexibility.

Reply:
I see! It makes sense